docs/language/ql-training-rst/cpp/bad-overflow-guard.rst

```diff
@@ -0,0 +1,228 @@
+
+Checking for overflow in C revisited
+====================================
+
+Which branch gets executed in this example? What is the value of ``result``?
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if (v + b < v) {
+      handle_error("overflow");
+  } else {
+      result = v + b;
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and ``result`` is set to zero.
+
+Checking for overflow in C revisited
+====================================
+
+Here is the example again, with the conversions made explicit:
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if ((int)v + (int)b < (int)v) {
+      handle_error("overflow");
+  } else {
+      result = (uint16_t)((int)v + (int)b);
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and result is set to zero.
+
+  Explanation:
+  - The two integer arguments to the addition, v and b, are promoted to 32 bit integers.
+  - The comparison (<) is also an arithmetic operation, therefore it will also be completed on 32 bit integers.
+  - This means that v + b < v will never be true, because v and b can hold at most 216.
+  - Therefore, the second branch is executed, but the result of the addition is stored into the result variable. Overflow will still occur as result is a 16 bit integer.
+
+This happens even though the overflow check passed!
+
+.. rst-class:: background2
+
+Developing a QL query
+=====================
+
+Finding bad overflow guards
+
+QL query: bad overflow guards
+=============================
+
+Let’s look for overflow guards of the form ``v + b < v``, using the classes
+``AddExpr``, ``Variable`` and ``RelationalOperation`` from the ``cpp`` library.
+
+.. rst-class:: build
+
+.. literalinclude:: ../query-examples/cpp/bad-overflow-guard-1.ql
+   :language: ql
+
+.. note::
+
+  - When performing `variant analysis <https://semmle.com/variant-analysis>`__, it is usually helpful to write a simple query that finds the simple syntactic pattern, before trying to go on to describe the cases where it goes wrong.
+  - In this case, we start by looking for all the *overflow* checks, before trying to refine the query to find all *bad overflow* checks.
+  - The select clause defines what this query is looking for:
+    - an ``AddExpr``: the expression that is being checked for overflow.
+    - a ``RelationalOperation``: the overflow comparison check.
+    - a ``Variable``: used as an argument to both the addition and comparison.
+  - The where part of the query ties these three QL variables together using `predicates <https://help.semmle.com/QL/ql-handbook/predicates.html>`__ defined in the `standard QL for C/C++ library <https://help.semmle.com/qldoc/cpp/>`__.
+
+QL query: bad overflow guards
```

Comment:
Not sure whether this is a good or a bad idea. Apparently you can use `nextslide` to have a set of slides with a single heading. It would be good from the maintainability POV, but less obvious what's going on to an editor/writer of the slides.